### PR TITLE
Xuzhu/chunk reader

### PR DESCRIPTION
**Pull request by xzhu1900**
This change includes the implementation of chunkDataSet. It contains three parts:
1. the worker threads to load chunks
2. the get_batch caller thread to retrieve batches
3. the thread management logic to keep 1. and 2. in sync.

### CODE REVIEW

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  virtual Batch read_chunk(size_t chunk_index) = 0;
+
+  /// Return the chunk count. A derived class needs to override this method.
+  virtual size_t get_chunk_count() = 0;
+
+  void initialize() {
+    /// After the data set is created, we need to call this method before
+    /// getting batch data. This method will use some infomation that is not yet
+    /// available in the constructor to configure chunk sampler and start
+    /// working thread.
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:48]: We need to get this method as a standalone PR. Hopefully, this is the only change we need for the dataset API.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+/// or part of a file in the case of a large text file split based on seek
+/// positions. ChunkDataSet extends the DataSet functionality to read an
+/// antire chunk at once.
+template <
+    typename Self,
+    typename Batch = std::vector<Example<>>,
+    typename ChunkSampler = samplers::SequentialSampler,
+    typename ExampleSampler = samplers::SequentialSampler>
+class ChunkDataSet : public BatchDataset<Self, Batch> {
+ public:
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:37]: Let's put the templates as typedef(using) for anybody to use later here.
E.g. using BatchType = Batch.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+  ChunkDataSet(
+      size_t prefetch_count,
+      ChunkSampler chunkSampler = samplers::SequentialSampler(0),
+      ExampleSampler exampleSampler = samplers::SequentialSampler(0))
+      : prefetch_count_(prefetch_count),
+        chunk_sampler_(chunkSampler),
+        example_sampler_(exampleSampler) {
+    AT_ASSERT(prefetch_count >= 1);
+
+    chunkData_.resize(prefetch_count_);
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:66]: chunk_data

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        batch_example_indices.push_back(
+            {current_reading_chunk,
+             example_position_,
+             example_position_ + copiableItemCount});
+
+        example_position_ += copiableItemCount;
+        count += copiableItemCount;
+      }
+
+      if (count < batch_size || example_position_ >= totalExampleInChunk) {
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:253]: can this ever be >, only equal is valid right?

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    }
+  }
+
+  struct RandExampleIndexRange {
+    size_t chunkIndex;
+    size_t startRandIndex;
+    size_t endRandIndex;
+  };
+
+  std::vector<RandExampleIndexRange> get_example_indices(size_t batch_size) {
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:226]: Say, we are on the last chunk, there is no more data to load. How does this method notify the caller that no more data is available?

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+      size_t prefetch_count,
+      ChunkSampler chunkSampler = samplers::SequentialSampler(0),
+      ExampleSampler exampleSampler = samplers::SequentialSampler(0))
+      : prefetch_count_(prefetch_count),
+        chunk_sampler_(chunkSampler),
+        example_sampler_(exampleSampler) {
+    AT_ASSERT(prefetch_count >= 1);
+
+    chunkData_.resize(prefetch_count_);
+    example_indices_.resize(prefetch_count_);
```
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:67]: Could we reverse the ordering. Here we have two flat vectors storing the indices. Could we instead use a struct ChunkIndices and hold them in a single vector. That way we can clear on the two pointer algorithm we discussed.

File: test/cpp/api/dataloader.cpp [hunk 7]
```diff
@@ -39,6 +39,71 @@ TEST(DataTest, DatasetCallsGetCorrectly) {
...
+
+  Batch read_chunk(size_t chunk_index) override
+  {
+    Batch batchData;
+    auto s = chunkSize[chunk_index];
+    batchData.resize(chunkSize[chunk_index]);
+    std::iota(batchData.begin(), batchData.end(), chunkSize[chunk_index]);
+    return batchData;
+  }
+  torch::optional<size_t> size() const override {
```
Comment (thiagocrepaldi) [inline test/cpp/api/dataloader.cpp:54]: As size() is not used by ChunkDataSet, it doesnt have to be overridden, right?

File: torch/csrc/api/include/torch/data/samplers/base.h [hunk 8]
```diff
@@ -28,6 +28,8 @@ class Sampler {
   /// Typically called before a new epoch.
   virtual void reset() = 0;
 
+  virtual void set_size(size_t new_size) = 0;
```
Comment (thiagocrepaldi) [inline torch/csrc/api/include/torch/data/samplers/base.h]: should it be `optional<size_t> new_size`?

File: test/cpp/api/dataloader.cpp (hunk 7 above)
Comment (xzhu1900) [reply test/cpp/api/dataloader.cpp:54]: size will be used internally in ChunkDataSet::Reset() (or initialized in this version). It is only known by the specific derived class. We need this number to reset chunk sampler before retrieving the chunk data.

**Pull request closed without merging**